proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,417 @@
+alternative tries to make the cluster automatically enter restart preparation 
+phase without human intervention.
+
+While getting humans out of the loop improves recovery speed, there are
+concerns about recovery gossip messages interfering with normal gossip 
+messages, and automatically start a new message in gossip seems risky.
+
+### Automatically reduce block production in an outage
+
+Right now we have vote-only mode, a validator will only pack vote transactions
+into new blocks if the tower distance (last_vote - local_root) is greater than
+400 slots.
+
+Unfortunately in the previous outages vote-only mode isn't enough to save the
+cluster. There are proposals of more aggressive block production reduction to
+save the cluster. For example, a leader could produce only one block in four
+consecutive slots allocated to it.
+
+However, this only solves the problem in specific type of outage, and it seems
+risky to aggressively reduce block production, so we are not proceeding with
+this proposal for now.
+
+## Detailed Design
+
+The new protocol tries to make all restarting validators get the same data
+blocks and the same set of last votes, so that they will with high probability
+converge on the same canonical fork and proceed.
+
+When the cluster is in need of a restart, we assume validators holding at least
+`RESTART_STAKE_THRESHOLD` percentage of stakes will enter the restart mode.
+Then the following steps will happen:
+
+1. The operator restarts the validator into the `wen restart phase` at boot,
+where it will not make new blocks or vote. The validator propagates its local
+voted fork information to all other validators in restart.
+
+2. While aggregating local vote information from all others in restart, the
+validator repairs all blocks which could potentially have been optimistically
+confirmed.
+
+3. After repair is complete, the validator counts votes on each fork and
+sends out local heaviest fork.
+
+4. Each validator counts if enough nodes can agree on one block (same slot and
+hash) to restart from:
+
+   1. If yes, proceed and restart
+
+   2. If no, print out what it thinks is wrong, halt and wait for human
+
+See each step explained in details below.
+
+### Wen restart phase
+
+1. **gossip last vote and ancestors on that fork**
+
+   The main goal of this step is to propagate most recent ancestors on the last
+   voted fork to all others in restart.
+
+   We use a new gossip message `RestartLastVotedForkSlots`, its fields are:
+
+   * `last_voted_slot`: `u64` the slot last voted, this also serves as
+   last_slot for the bit vector.
+   * `last_voted_hash`: `Hash` the bank hash of the slot last voted slot.
+   * `ancestors`: `BitVec<u8>` compressed bit vector representing the slots on
+   sender's last voted fork. the most significant bit is always
+   `last_voted_slot`, least significant bit is `last_voted_slot-81000`.
+
+   The number of ancestor slots sent is hard coded at 81000, because that's
```

Comment:
A quick justification for 81000 would be nice, to help the reader with context. I don't recall myself either! But I assume we're trying to limit the max size of the crds message?

Reply:
I remember that we set it at 9 hours because we assumed that most validator admins would wake up and join the choir in 9 hours. 

In reality if the cluster gets start, new blocks would be confirmed and voted on very slowly, so very likely 81k slots is overkill.

Reply:
Added motivation to the description.

Reply:
Got it. So there's no underlying reason why 81k couldn't be larger beyond this? Like if we said 24 hours worth of slots is the very worst case, is there any downside to doing so if in reality all future restarts are handled in less than 9 hours?

Reply:
Also, 81k is already larger than one UDP packet, so it needs to be chopped into chunks. So 24 hours is not much worse, just more packets.